### PR TITLE
migrates enum attribute size to 255

### PR DESCRIPTION
**Pull request by fanatic75**
## What does this PR do?

Migrates enum attribute size to 255 for bug fix https://github.com/appwrite/appwrite/issues/5595

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [ ] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/init.php [hunk 1]
```diff
@@ -767,6 +767,25 @@ function (mixed $value) {
...
+           PDO::ATTR_TIMEOUT => 3, // Seconds
+           PDO::ATTR_PERSISTENT => true,
+           PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
+           PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
+           PDO::ATTR_EMULATE_PREPARES => true,
+           PDO::ATTR_STRINGIFY_FETCHES => true,
+       ));
+
+       return $pdo;
+});
```
Comment (christyjacob4) [inline app/init.php:788]: why do we need this? you can use the db connection from the database pool directly

File: src/Appwrite/Migration/Version/V19.php [hunk 2]
```diff
@@ -640,6 +646,21 @@ private function getFunctionCommands(Document $function): string
...
+    private function migrateEnumAttributeSize(): void
+    {
+        \Co\run(function () {
+            foreach ($this->documentsIterator('attributes') as $attribute) {
+                go(function ($attribute) {
+                    $attribute->setAttribute('size', Database::LENGTH_KEY);
+                    $this->projectDB->updateDocument('attributes', $attribute->getId(), $attribute);
+                    $databaseInternalId = $attribute->getAttribute('databaseInternalId');
+                    $collectionInternalId = $attribute->getAttribute('collectionInternalId');
+                    $this->changeAttributeInternalType('database_' . $databaseInternalId . '_collection_' . $collectionInternalId, $attribute->getAttribute('key'), 'varchar(255)');
```
Comment (stnguyen90) [inline src/Appwrite/Migration/Version/V19.php:658]: It would be good to avoid adapter specific things when possible. Can we use [`$this->projectDB->updateAttribute()`](https://github.com/utopia-php/database/blob/60a61d00570bb66e2277be9c88b947279c729623/src/Database/Database.php#L1310) instead?

File: src/Appwrite/Migration/Version/V19.php [hunk 3]
```diff
@@ -640,6 +646,21 @@ private function getFunctionCommands(Document $function): string
         return $commands;
     }
 
+    private function migrateEnumAttributeSize(): void
+    {
+        \Co\run(function () {
+            foreach ($this->documentsIterator('attributes') as $attribute) {
+                go(function ($attribute) {
```
Comment (stnguyen90) [inline src/Appwrite/Migration/Version/V19.php:653]: Considering we have to update the collection record in the metadata table, it might be dangerous to do the updates concurrently.

File: src/Appwrite/Migration/Version/V19.php [hunk 4]
```diff
@@ -640,6 +646,21 @@ private function getFunctionCommands(Document $function): string
...
+                    $attribute->setAttribute('size', Database::LENGTH_KEY);
```
Comment (stnguyen90) [inline src/Appwrite/Migration/Version/V19.php:654]: We should check the current size and make sure we're not making it smaller than what it was before.

File: app/init.php (hunk 1 above)
Comment (fanatic75) [reply app/init.php:788]: We are missing setting up PDO in migrate file. This is required for using PDO during migrations

File: src/Appwrite/Migration/Version/V19.php (hunk 3 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php:653]: We are not updating metadata table here. We are updating attributes table and collection table

File: src/Appwrite/Migration/Version/V19.php (hunk 2 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php:658]: what is the adapter specific thing here?

File: src/Appwrite/Migration/Version/V19.php (hunk 3 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php]: removed co routines.

File: src/Appwrite/Migration/Version/V19.php (hunk 4 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php]: Added condition for checking current size, although something else will be needed to solve the bug. I think we should truncate the value and update the value to maximum 255 length size. 
cc @abnegate

File: src/Appwrite/Migration/Version/V19.php (hunk 2 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php]: Changed to updateAttribute

**Review by wess — APPROVED** [review]

File: app/init.php (hunk 1 above)
Comment (abnegate) [reply app/init.php:788]: We can remove this now we are using `updateAttribute`, we should try keep moving towards using only the db lib functions :+1:

File: src/Appwrite/Platform/Tasks/Migrate.php [hunk 5]
```diff
@@ -29,7 +30,8 @@ public function __construct()
             ->inject('cache')
             ->inject('dbForConsole')
             ->inject('getProjectDB')
-            ->callback(fn ($version, $cache, $dbForConsole, $getProjectDB) => $this->action($version, $cache, $dbForConsole, $getProjectDB));
+            ->inject('register')
```
Comment (abnegate) [inline src/Appwrite/Platform/Tasks/Migrate.php:33]: We can revert the changes here

File: src/Appwrite/Migration/Version/V19.php (hunk 4 above)
Comment (abnegate) [reply src/Appwrite/Migration/Version/V19.php]: Actually I think it should be okay to ignore ones that are longer than 255 already 🤔 any developers in that scenario just won't be able to add any _more_ that are longer than 255

Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php]: They will be able to update those enum attributes with length greater than 255. Update endpoint doesn't restrict size.

File: app/init.php (hunk 1 above)
Comment (fanatic75) [reply app/init.php:788]: keeping it here @abnegate  for migration files that require PDO

File: src/Appwrite/Migration/Version/V19.php (hunk 4 above)
Comment (fanatic75) [reply src/Appwrite/Migration/Version/V19.php]: Okay. resolving this then

File: src/Appwrite/Platform/Tasks/Migrate.php (hunk 5 above)
Comment (fanatic75) [reply src/Appwrite/Platform/Tasks/Migrate.php:33]: Keeping it as it is required for setting PDO

File: src/Appwrite/Migration/Migration.php [hunk 6]
```diff
@@ -195,17 +195,23 @@ public function forEachDocument(callable $callback): void
...
         $sum = 0;
         $nextDocument = null;
         $collectionCount = $this->projectDB->count($collectionId);
+        $queries[] = Query::limit($this->limit);
 
         do {
-            $queries = [Query::limit($this->limit)];
             if ($nextDocument !== null) {
-                $queries[] = Query::cursorAfter($nextDocument);
+                $lastQuery = $queries[\count($queries) - 1];
```
Comment (abnegate) [inline src/Appwrite/Migration/Migration.php:207]: What if the `cursorAfter` is not the last query passed but the first of multiple?

File: src/Appwrite/Migration/Version/V19.php [hunk 7]
```diff
@@ -640,6 +646,17 @@ private function getFunctionCommands(Document $function): string
         return $commands;
     }
 
+    private function migrateEnumAttributeSize(): void
+    {
+        foreach ($this->documentsIterator('attributes', [Query::equal('format', ['enum']), Query::lessThan('size', Database::LENGTH_KEY)]) as $attribute) {
```
Comment (abnegate) [inline src/Appwrite/Migration/Version/V19.php:651]: Let's line break the queries for readability:

```suggestion
        foreach ($this->documentsIterator('attributes', [
            Query::equal('format', ['enum']), 
            Query::lessThan('size', Database::LENGTH_KEY)
        ]) as $attribute) {
```

**Review by abnegate — COMMENTED** [review]
Looking good, please run a manual test to make sure it's working as expected too

**Review by abnegate — APPROVED** [review]
